core/src/main/scalajvm/scalapb/zio_grpc/server/ZServerCallHandler.scala

```diff
@@ -108,30 +111,59 @@ object ZServerCallHandler {
       call: ZServerCall[Res],
       stream: ZStream[Any, Status, Res]
   ): ZIO[Any, Status, Unit] = {
-    def innerLoop(queue: Dequeue[Exit[Option[Status], Res]]): ZIO[Any, Status, Boolean] =
-      queue.take
-        .flatMap {
-          case Exit.Success(res)   => call.sendMessage(res).as(true)
-          case Exit.Failure(cause) =>
-            cause.failureOrCause match {
-              case Left(Some(status)) => ZIO.fail(status)
-              case Left(None)         => ZIO.succeed(false)
-              case Right(cause)       => ZIO.failCause(cause)
+    def takeFromQueue(queue: Dequeue[Exit[Option[Status], Res]]): ZIO[Any, Status, Unit] =
+      queue.takeAll.flatMap(takeFromCache(_, queue))
+
+    def takeFromCache(
+        xs: Chunk[Exit[Option[Status], Res]],
+        queue: Dequeue[Exit[Option[Status], Res]]
+    ): ZIO[Any, Status, Unit] =
+      ZIO.succeed {
+        @tailrec def innerLoop(loop: Boolean, i: Int): IO[Status, Unit] =
+          if (loop) {
+            xs(i) match {
+              case Failure(cause) =>
+                cause.failureOrCause match {
+                  case Left(Some(status)) =>
+                    ZIO.fail(status)
+                  case Left(None)         =>
+                    ZIO.unit
+                  case Right(cause)       =>
+                    ZIO.failCause(cause)
+                }
+              case Success(value) =>
+                call.call.sendMessage(value)
```

Comment:
It looks like it's doing a direct call to a side-effect from here. So while `innerLoop` seems pure (returns an `IO`), it does networking on the side.